### PR TITLE
Pass Through View Tests for Sync Gateway

### LINKED ISSUE #241: Add functional test to exercise pass-thru view query behavior
Sync Gateway has the feature to pass-thru calls to couchbase server views.  It's a very powerful and useful feature.  I don't know if this is being tested at all by our functional tests (or integration tests).
- [Open issue: Pass-thru view query is failing](https://github.com/couchbase/sync_gateway/issues/1493)
- [Documentation on Sync Gateway views](http://developer.couchbase.com/documentation/mobile/1.1.0/develop/guides/sync-gateway/accessing-cb-views/creating-and-using-a-view/index.html)

### LINKED ISSUE #276: Add tests for pass through views
Make sure this is verified against Couchbase Server back sync_gateways as walrus does not behave the same. 
- [x] Test multiple views in the same design doc
- [x] Test admin / user (requires enabling experimental features)

### LINKED ISSUE #1044: Add tests that exercise SG Principle views
- principles

### PR DESCRIPTION
**Pull request by sethrosetter**
#### Fixes #.

- [x] Ran `flake8`

#### Changes proposed in this pull request:

- Fixes #1044 
- Fixes #276 
- Fixes #241 
- New attachment api
- Add unit tests for keywords/attachment.py and keywords/document.py
- Add a few attachment generator functions
- Add pass through view test
- Add `get_roles` and `get_users` to MobileRestClient.py
- Support pass through views for sync gateway
- Add script to run framework unit tests without deps 

Functional test run (In Progress) - http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-cc/321/

### CODE REVIEW

**Comment by sethrosetter** [comment]
@tleyden, per out conversation yesterday, you said you would be interested in looking at this.
@adamcfraser, if you can look at the test case and see if what you think
@raghusarangapani, please review carefully. This is a pretty big change.

**Comment by sethrosetter** [comment]
Jenkins jobs
- http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-cc/321/
- http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-base-di/294/
- http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-cc/249/ 
- http://uberjenkins.sc.couchbase.com/job/cen7-sync-gateway-functional-tests-topology-specific-di/226/

File: keywords/MobileRestClient.py [hunk 1]
```diff
@@ -1715,15 +1734,33 @@ def add_design_doc(self, url, db, name, doc):
...
         resp.raise_for_status()
 
-        resp_obj = resp.json()
-
-        return resp_obj["id"]
+        # Only return a response if adding to the listener
+        # Sync Gateway does not return a response
+        if self.get_server_type(url) == ServerType.listener:
+            resp_obj = resp.json()
+            return resp_obj["id"]
```
Comment (raghusarangapani) [inline keywords/MobileRestClient.py]: Is it possible for a SG test to call this function? If so, will it panic upon no return value?

File: keywords/attachment.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+
+def generate_2_png_100_100():
+    att_one_list = generate_png(100, 100)
+    att_two_list = generate_png(100, 100)
+    return att_one_list + att_two_list
+
+
+def generate_png(width, height):
+    """ Generates a noise rgb images for attachment testing. """
```
Comment (raghusarangapani) [inline keywords/attachment.py]: Very cool! :)

File: testsuites/syncgateway/functional/tests/test_user_views.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    users = client.get_users(url=sg_admin_url, db=sg_db)
+
+    # These are defined in the config
+    assert len(users) == 2 and "seth" in users and "raghu" in users
+
+    # Issue GET /_role to exercise principal views
+    roles = client.get_roles(url=sg_admin_url, db=sg_db)
+
+    # These are defined in the config
+    assert len(roles) == 2 and "Scientist" in roles and "Researcher" in roles
```
Comment (raghusarangapani) [inline testsuites/syncgateway/functional/tests/test_user_views.py]: It is also possible to verify user to role mapping?

File: keywords/MobileRestClient.py (hunk 1 above)
Comment (sethrosetter) [inline keywords/MobileRestClient.py]: SG has no return value. This is why this was added. I talked to @adamcfraser about the inconsistency and filed a ticket - https://github.com/couchbase/sync_gateway/issues/2413

**Pull request merged**